### PR TITLE
Single Number Challenge

### PR DESCRIPTION
**Pull request by aoyeoku**
### Challenge Description
Given a non-empty array of integers nums, every element appears twice except for one. Find that single one.

### Example 1:
Input: nums = [2,2,1]

Output: 1

### Example 2:
Input: nums = [4,1,2,1,2]

Output: 4

### Example 3:
Input: nums = [1]

Output: 1

Constraints:

1 <= nums.length <= 3 * 104
-3 * 104 <= nums[i] <= 3 * 104
Each element in the array appears twice except for one element which appears only once.

Helpful Resources:
[AlgoMonster](https://algo.monster/liteproblems/136)
[DEV Community Discussion](https://dev.to/melguachun/leetcode-136-single-number-200d)

### Files

- [ ] The file name describes the function's behavior
- [ ] There is a module docstring in the function file
- [ ] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [ ] There is a module docstring in the tests file

### Unit Tests

- [ ] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Every unit test has
  - [ ] A helpful name
  - [ ] A clear docstring
  - [ ] Only one assertion
  - [ ] There is no logic in the unit test
- [ ] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [ ] The function's behavior is described
- The function's arguments are described:
  - [ ] Type
  - [ ] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [ ] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [ ] The function's name describes it's behavior
- [ ] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [ ] The function has correct type annotations
- [ ] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [ ] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [ ] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in any docstrings or tests
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [ ] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [ ] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [ ] Variables are named with snake_case
- [ ] Variable names are clear and helpful
- [ ] The code follows the strategy as simply as possible
- [ ] The implementation is as simple as possible given the strategy
- [ ] There are no commented lines of code
- [ ] There are no `print` statements anywhere
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/function_single_number_challenge.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+in a list where every element appears exactly twice except for one
+element that appears only once.
+
+
+Created on 2024-01-11
+@author: Abbiihjah Oyeoku
+"""
+
+
+def single_number(nums):
```
Comment (ArwaAbdelkhalik) [inline solutions/function_single_number_challenge.py:17]: The function has no time annotation

File: solutions/function_single_number_challenge.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    """
+    Finds the single number in a list where every element appears twice except for one.
+
+    This function uses the XOR operation to efficiently determine the unique number
+    that appears only once in the input list. The XOR operation ensures that numbers
+    appearing in pairs cancel each other out, leaving the single number.
+
+    Args:
```
Comment (ArwaAbdelkhalik) [inline solutions/function_single_number_challenge.py:25]: Other assumptions: Does not specify assumptions like whether lisits with empty strings or spaces.

File: solutions/function_single_number_challenge.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        4
+        >>> single_number([1])
+        1
+    """
+    if not nums:  # Check if the list is empty
+        raise ValueError("Input list cannot be empty")
+    # Initialize the unique number...
+    uniq_num = 0
+    # TRaverse all elements through the loop...
+    for idx in nums:
```
Comment (ArwaAbdelkhalik) [inline solutions/function_single_number_challenge.py:45]: comments are clear and describe the strategy

File: solutions/tests/test_single_number_challenge.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    Unit test class for testing the `single_number` function.
+    """
+
+    # Standard Tests
+    def test_standard_case(self):
+        """
+        Standard: Test with typical input.
+        """
+        self.assertEqual(single_number([2, 2, 1]), 1, "Failed on input [2, 2, 1]")
+        self.assertEqual(
```
Comment (ArwaAbdelkhalik) [inline solutions/tests/test_single_number_challenge.py:35]: every unit test has one assertion

File: solutions/function_single_number_challenge.py (hunk 1 above)
Comment (aoyeoku) [reply solutions/function_single_number_challenge.py:17]: Hi Arwa, I was not aware that the functions required time annotations.

File: solutions/function_single_number_challenge.py (hunk 2 above)
Comment (aoyeoku) [reply solutions/function_single_number_challenge.py:25]: I have an assertion for the empty list.

File: solutions/function_single_number_challenge.py (hunk 3 above)
Comment (aoyeoku) [reply solutions/function_single_number_challenge.py:45]: Noted, thank you.

File: solutions/function_single_number_challenge.py (hunk 1 above)
Comment (misik-eng) [reply solutions/function_single_number_challenge.py:17]: I did not know about time annotations and don’t think we strictly need them for this project, but I see that you added them. Thank you!

File: solutions/function_single_number_challenge.py (hunk 2 above)
Comment (misik-eng) [reply solutions/function_single_number_challenge.py:25]: I think it is enough to have an empty list, and it is not strictly necessary to have assertions or assumptions for empty strings or spaces.

File: solutions/function_single_number_challenge.py (hunk 3 above)
Comment (misik-eng) [reply solutions/function_single_number_challenge.py:53]: Yep, maybe there could be more comments, but I think this version is also good.

**Review by misik-eng — APPROVED** [review]
I checked your pr and see taht you made most of the changes requested. Thank you.
 I approve your pr.
 
 ### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [X] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

## Strategy

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in any docstrings or tests
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible
